File: go/examples/filtering.go
//go:build ignore
// +build ignore

package main

import (
	"fmt"
	"time"

	tracing "github.com/LeDuyViet/quicktrace/go"
)

func simulateVariousOperations(tracer *tracing.Tracer) {
	// Ultra fast operations
	tracer.Span("Ultra fast operation 1")
	time.Sleep(500 * time.Microsecond)

	tracer.Span("Ultra fast operation 2")
	time.Sleep(800 * time.Microsecond)

	// Fast operations
	tracer.Span("Fast validation")
	time.Sleep(5 * time.Millisecond)

	tracer.Span("Quick lookup")
	time.Sleep(8 * time.Millisecond)

	// Medium operations
	tracer.Span("Medium processing")
	time.Sleep(45 * time.Millisecond)

	tracer.Span("Similar processing")
	time.Sleep(48 * time.Millisecond) // Similar to above

	tracer.Span("Another similar task")
	time.Sleep(44 * time.Millisecond) // Also similar

	// Slow operations
	tracer.Span("Slow database query")
	time.Sleep(150 * time.Millisecond)

	tracer.Span("Complex computation")
	time.Sleep(200 * time.Millisecond)

	// Very slow operation
	tracer.Span("Very slow external API call")
	time.Sleep(800 * time.Millisecond)
}

func main() {
	fmt.Println("🎯 Smart Filtering Examples")
	fmt.Println("===========================")

	// Example 1: No filtering (show all)
	fmt.Println("\n1️⃣ No Filtering (Show All Operations)")
	fmt.Println("-------------------------------------")
	tracer1 := tracing.NewSimpleTracer("Complete Trace",
		tracing.WithOutputStyle(tracing.StyleDetailed))
	simulateVariousOperations(tracer1)
	tracer1.End()

	time.Sleep(1 * time.Second)

	// Example 2: Hide ultra fast operations
	fmt.Println("\n2️⃣ Hide Ultra Fast Operations (< 2ms)")
	fmt.Println("--------------------------------------")
	tracer2 := tracing.NewSimpleTracer("Filtered - No Ultra Fast",
		tracing.WithOutputStyle(tracing.StyleDetailed),
		tracing.WithHideUltraFast(2*time.Millisecond))
	simulateVariousOperations(tracer2)
	tracer2.End()

	time.Sleep(1 * time.Second)

	// Example 3: Show only slow operations
	fmt.Println("\n3️⃣ Show Only Slow Operations (>= 100ms)")
	fmt.Println("----------------------------------------")
	tracer3 := tracing.NewSimpleTracer("Slow Operations Only",
		tracing.WithOutputStyle(tracing.StyleDetailed),
		tracing.WithShowSlowOnly(100*time.Millisecond))
	simulateVariousOperations(tracer3)
	tracer3.End()

	time.Sleep(1 * time.Second)

	// Example 4: Group similar operations
	fmt.Println("\n4️⃣ Group Similar Operations (±10ms threshold)")
	fmt.Println("----------------------------------------------")
	tracer4 := tracing.NewSimpleTracer("Grouped Similar",
		tracing.WithOutputStyle(tracing.StyleDetailed),
		tracing.WithGroupSimilar(10*time.Millisecond))
	simulateVariousOperations(tracer4)
	tracer4.End()

	time.Sleep(1 * time.Second)

	// Example 5: Combined smart filtering
	fmt.Println("\n5️⃣ Combined Smart Filtering")
	fmt.Println("----------------------------")
	tracer5 := tracing.NewSimpleTracer("Smart Filtered",
		tracing.WithOutputStyle(tracing.StyleDetailed),
		tracing.WithSmartFilter(
			50*time.Millisecond,  // Show slow >= 50ms
			2*time.Millisecond,   // Hide ultra fast < 2ms
			15*time.Millisecond)) // Group similar ±15ms
	simulateVariousOperations(tracer5)
	tracer5.End()
}

File: js/examples/filtering.js
const { Tracer } = require('../tracer.js');

async function simulateVariousOperations(tracer) {
    // Ultra fast operations
    await new Promise(resolve => setTimeout(resolve, 0.5));
    tracer.span('Ultra fast operation 1');

    await new Promise(resolve => setTimeout(resolve, 0.8));
    tracer.span('Ultra fast operation 2');

    // Fast operations
    await new Promise(resolve => setTimeout(resolve, 5));
    tracer.span('Fast validation');

    await new Promise(resolve => setTimeout(resolve, 8));
    tracer.span('Quick lookup');

    // Medium operations
    await new Promise(resolve => setTimeout(resolve, 45));
    tracer.span('Medium processing');

    await new Promise(resolve => setTimeout(resolve, 48)); // Similar to above
    tracer.span('Similar processing');

    await new Promise(resolve => setTimeout(resolve, 44)); // Also similar
    tracer.span('Another similar task');

    // Slow operations
    await new Promise(resolve => setTimeout(resolve, 150));
    tracer.span('Slow database query');

    await new Promise(resolve => setTimeout(resolve, 200));
    tracer.span('Complex computation');

    // Very slow operation
    await new Promise(resolve => setTimeout(resolve, 800));
    tracer.span('Very slow external API call');
}

async function main() {
    console.log('🎯 Smart Filtering Examples');
    console.log('===========================');

    // Example 1: No filtering (show all)
    console.log('\n1️⃣ No Filtering (Show All Operations)');
    console.log('-------------------------------------');
    const tracer1 = new Tracer('Complete Trace', {
        style: 'detailed'
    });
    await simulateVariousOperations(tracer1);
    tracer1.end();

    await new Promise(resolve => setTimeout(resolve, 1000));

    // Example 2: Hide ultra fast operations
    console.log('\n2️⃣ Hide Ultra Fast Operations (< 2ms)');
    console.log('--------------------------------------');
    const tracer2 = new Tracer('Filtered - No Ultra Fast', {
        style: 'detailed',
        hideUltraFast: 2 // ms
    });
    await simulateVariousOperations(tracer2);
    tracer2.end();

    await new Promise(resolve => setTimeout(resolve, 1000));

    // Example 3: Show only slow operations
    console.log('\n3️⃣ Show Only Slow Operations (>= 100ms)');
    console.log('----------------------------------------');
    const tracer3 = new Tracer('Slow Operations Only', {
        style: 'detailed',
        showSlowOnly: 100 // ms
    });
    await simulateVariousOperations(tracer3);
    tracer3.end();

    await new Promise(resolve => setTimeout(resolve, 1000));

    // Example 4: Group similar operations
    console.log('\n4️⃣ Group Similar Operations (±10ms threshold)');
    console.log('----------------------------------------------');
    const tracer4 = new Tracer('Grouped Similar', {
        style: 'detailed',
        groupSimilar: 10 // ms
    });
    await simulateVariousOperations(tracer4);
    tracer4.end();

    await new Promise(resolve => setTimeout(resolve, 1000));

    // Example 5: Combined smart filtering
    console.log('\n5️⃣ Combined Smart Filtering');
    console.log('----------------------------');
    const tracer5 = new Tracer('Smart Filtered', {
        style: 'detailed',
        showSlowOnly: 50,      // Show slow >= 50ms
        hideUltraFast: 2,      // Hide ultra fast < 2ms
        groupSimilar: 15,      // Group similar ±15ms
        minTotalDuration: 100  // Only show if total >= 100ms
    });
    await simulateVariousOperations(tracer5);
    tracer5.end();
}

main().catch(console.error); 